### ISSUE TITLE
'iCalendar not well-formed' error when adding CalDAV calendar

### ISSUE THREAD

**Issue by ashjrl** [marcoancona/MineTime#139]
I am using MineTime 1.6.1 - this problem has been happening for a while now though, it's not new.

I have some work related calendars which I am trying to set up - the mail server is Kerio Connect ( https://www.gfi.com/products-and-solutions/email-and-messaging-solutions/kerio-connect )

I do have other CalDAV calendars added which are working fine.

When I add the account, it works and all calendars are displayed.

However, no events are displayed and a black box is constantly displayed at the bottom of the calendar saying "iCalendar not well-formed'.

Additionally, my laptop fan kicks in and the MineTime application becomes very slow and sluggish - it takes me a few minutes to successfully delete the calendar before things go back to normal.

I have never had an issue before adding these calendars to other calendar clients.

From the log file:

> 2019-09-05T09:36:47.681Z - info: Creating CalDAV account...
2019-09-05T09:37:01.730Z - warn: iCalendar parsing error message=invalid ical body. component began but did not end, name=ParserError, stack=    at new ParserError (/opt/MineTime/resources/app.asar/node_modules/ical.js/build/ical.js:1702:13)
    at Object.parser [as parse] (/opt/MineTime/resources/app.asar/node_modules/ical.js/build/ical.js:1740:13)
    at icalToEvents (/opt/MineTime/resources/app.asar/dist/main/helpers/ical.js:1:1978)
    at storeRemoteUpdateLocally (/opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:2976)
    at /opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:7004
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:3110:16
    at replenish (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1011:17)
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1016:9
    at eachLimit$1 (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:3196:24)
    at Object.<anonymous> (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1046:16)
    at /opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:6840
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:3880:24
    at replenish (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1011:17)
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1016:9
    at eachOfLimit (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1041:24)
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1046:16
2019-09-05T09:37:01.752Z - warn: iCalendar parsing error message=invalid ical body. component began but did not end, name=ParserError, stack=    at new ParserError (/opt/MineTime/resources/app.asar/node_modules/ical.js/build/ical.js:1702:13)
    at Object.parser [as parse] (/opt/MineTime/resources/app.asar/node_modules/ical.js/build/ical.js:1740:13)
    at icalToEvents (/opt/MineTime/resources/app.asar/dist/main/helpers/ical.js:1:1978)
    at storeRemoteUpdateLocally (/opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:2976)
    at /opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:7004
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:3110:16
    at replenish (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1011:17)
    at iterateeCallback (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:995:17)
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:969:16
    at /opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:4135
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:473:16
    at replenish (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1006:25)
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1016:9
    at eachLimit$1 (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:3196:24)
    at Object.<anonymous> (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1046:16)
    at /opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:4009
2019-09-05T09:37:01.760Z - warn: iCalendar parsing error message=invalid ical body. component began but did not end, name=ParserError, stack=    at new ParserError (/opt/MineTime/resources/app.asar/node_modules/ical.js/build/ical.js:1702:13)
    at Object.parser [as parse] (/opt/MineTime/resources/app.asar/node_modules/ical.js/build/ical.js:1740:13)
    at icalToEvents (/opt/MineTime/resources/app.asar/dist/main/helpers/ical.js:1:1978)
    at storeRemoteUpdateLocally (/opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:2976)
    at /opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:7004
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:3110:16
    at replenish (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1011:17)
    at iterateeCallback (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:995:17)
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:969:16
    at /opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:4135
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:473:16
    at replenish (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1006:25)
    at /opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1016:9
    at eachLimit$1 (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:3196:24)
    at Object.<anonymous> (/opt/MineTime/resources/app.asar/node_modules/async/dist/async.js:1046:16)
    at /opt/MineTime/resources/app.asar/dist/main/connectors/calendars/dav/dav.js:1:4009

And many more of the same error logged..

**Comment by marcoancona** [comment]
Thanks for reporting. Apparently, there are some edge cases where CalDAV still fails. 
I have some questions that could help to find the issue:
- are the other caldav calendars (those that work) from the same Kerio account?
- on which other calendar client does it work (so we can compare the behavior)?

**Comment by ashjrl** [comment]
Thank you for responding so quickly.

No they are not on the same Kerio server - one that I can confirm that works is a calendar installed on a Roundcube webmail installation ( https://roundcube.net/ ) using the following Calendar plugin: https://roundcubeplus.com/plugins/calendar

One client that I can confirm works with our Kerio calendars is the Mozilla Thunderbird mail client ( https://www.thunderbird.net ) using the officially provided Lightning calendar plugin: https://www.thunderbird.net/en-US/calendar/

If it would be helpful I can provide you with a test account on both platforms (Kerio and Roundcube). If you do want this, please let me know how I can privately provide these details.

Ashley

**Comment by marcoancona** [comment]
Test accounts would be very useful. Please email us connect [at] minetime.ai

**Comment by marcoancona** [comment]
Good news, we managed to reproduce the problem. A fix will come with the next release.

**Comment by ashjrl** [comment]
Thank you very much - sorry I had not yet provided test accounts, I am away on vacation this week, back Monday. I look forward to the next release.

**Comment by ashjrl** [comment]
I have just upgraded to 1.6.2 - the problem appears to be mostly fixed, I have successfully added my calendars and events are appearing. Thank you very much! I have had this issue for a while and convinced myself that I should take part in making MineTime better and report it and I'm glad I did - the whole process was very quick and you were very helpful / responsive.

The black box saying 'iCalendar not well-formed' did briefly appear so I suspect there may still be an issue parsing a single event or some events.. I will try to see if I can identify an event that is missing from or looks corrupted in MineTime and let you know, I am yet to identify that, will keep this issue closed for now.
